### ISSUE TITLE
undefined reference to symbol 'floor@@GLIBC_2.2.5'

### ISSUE THREAD

**Issue by Alex20129** [ec-/Quake3e#135]
i try to compile Quake3e under Ubuntu 20.04.3 and get an error. i do:

```
alex@Alex-WS-X470GP:~/Quake3e/build$ cmake ..
-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is GNU 9.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found SDL2: /usr/lib/x86_64-linux-gnu/libSDL2.so (found version "2.0.10") 
-- Found SDL2main: /usr/lib/x86_64-linux-gnu/libSDL2main.a (found version "2.0.10") 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/alex/Quake3e/build
```
it looks Ok. Then i do:
```
alex@Alex-WS-X470GP:~/Quake3e/build$  make
```
compilation process going normally till the some point
```
[ 72%] Building C object CMakeFiles/quake3e.dir/code/qcommon/vm_x86.c.o
[ 72%] Building C object CMakeFiles/quake3e.dir/code/unix/unix_main.c.o
[ 72%] Building C object CMakeFiles/quake3e.dir/code/unix/unix_shared.c.o
[ 73%] Building C object CMakeFiles/quake3e.dir/code/unix/linux_signals.c.o
[ 73%] Linking C executable quake3e
/usr/bin/ld: libclient.a(cl_ui.c.o): undefined reference to symbol 'floor@@GLIBC_2.2.5'
/usr/bin/ld: /lib/x86_64-linux-gnu/libm.so.6: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/quake3e.dir/build.make:139: quake3e] Error 1
make[1]: *** [CMakeFiles/Makefile2:134: CMakeFiles/quake3e.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
```
what i'm doing wrong?

**Comment by ensiform** [comment]
Do you have any better luck using the makefile instead? Looks like an incompatibility with your glibc version...

**Comment by Alex20129** [comment]
> Do you have any better luck using the makefile instead? Looks like an incompatibility with your glibc version...

i don't understand, what's the point of using makefile in the CMake project. what exactly i need to do?

**Comment by ensiform** [comment]
The cmake isn't the primary way to build the project, it's a basic cmake file added from a PR.

**Comment by Alex20129** [comment]
i added the libm at line 142
```
target_link_libraries(quake3e client botlib m)
```
new error:
```
unix_shared.c.o: undefined reference to symbol 'dlclose@@GLIBC_2.2.5'
```
so, i added the libdl
```
target_link_libraries(quake3e client botlib m dl)
```
and then i get a new errors:
```
[ 71%] Built target qsdl
[ 72%] Building C object CMakeFiles/quake3e.dir/code/qcommon/vm_x86.c.o
[ 72%] Building C object CMakeFiles/quake3e.dir/code/unix/unix_main.c.o
[ 72%] Building C object CMakeFiles/quake3e.dir/code/unix/unix_shared.c.o
[ 73%] Building C object CMakeFiles/quake3e.dir/code/unix/linux_signals.c.o
[ 73%] Linking C executable quake3e
/usr/bin/ld: libserver.a(sv_bot.c.o): in function `SV_BotInitBotLib':
/home/alex/Quake3e/code/server/sv_bot.c:579: undefined reference to `GetBotLibAPI'
/usr/bin/ld: librenderer_common.a(tr_image_png.c.o): in function `DecompressIDATs':
/home/alex/Quake3e/code/renderercommon/tr_image_png.c:717: undefined reference to `puff'
/usr/bin/ld: /home/alex/Quake3e/code/renderercommon/tr_image_png.c:749: undefined reference to `puff'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/quake3e.dir/build.make:139: quake3e] Error 1
make[1]: *** [CMakeFiles/Makefile2:134: CMakeFiles/quake3e.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
02:30:24: The process "/usr/bin/cmake" exited with code 2.
Error while building/deploying project quake3e (kit: Desktop)
When executing step "CMake Build"
```
actually, i want to fix it or, at least help to fix it

**Comment by Alex20129** [comment]
i did it. just a few fixes in CMakeLists.txt file.
i also added the linker option
```
TARGET_LINK_OPTIONS(quake3e PRIVATE -no-pie)
```
to get an _executable_ instead of _shared library_. don't know if i did all the things right, but, here is the result
```
CMAKE_MINIMUM_REQUIRED(VERSION 3.10)
PROJECT(quake3e)

OPTION(USE_SDL "" ON)
OPTION(USE_CURL "" ON)
OPTION(USE_LOCAL_HEADERS "" ON)
OPTION(USE_VULKAN "" ON)
OPTION(USE_SYSTEM_JPEG "" OFF)
OPTION(USE_RENDERER_DLOPEN "" OFF)

SET(CMAKE_MODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/cmake_modules)

AUX_SOURCE_DIRECTORY(code/libjpeg JPEG_SRCS)
ADD_LIBRARY(jpeg ${JPEG_SRCS})


AUX_SOURCE_DIRECTORY(code/server SERVER_SRCS)
list(REMOVE_ITEM SERVER_SRCS code/server/sv_rankings.c)

ADD_LIBRARY(server ${SERVER_SRCS})
TARGET_LINK_LIBRARIES(server botlib qcommon)

ADD_LIBRARY(server_ded ${SERVER_SRCS})
TARGET_LINK_LIBRARIES(server_ded botlib qcommon)
TARGET_COMPILE_DEFINITIONS(server_ded PUBLIC -DDEDICATED)


AUX_SOURCE_DIRECTORY(code/qcommon QCOMMON_SRCS)
list(FILTER QCOMMON_SRCS EXCLUDE REGEX ".*vm_[alx].*.c")

ADD_LIBRARY(qcommon ${QCOMMON_SRCS} ${VM_SRC})
TARGET_LINK_LIBRARIES(qcommon server m dl)

ADD_LIBRARY(qcommon_ded ${QCOMMON_SRCS} ${VM_SRC})
TARGET_LINK_LIBRARIES(qcommon_ded server_ded m dl)


AUX_SOURCE_DIRECTORY(code/botlib BOTLIB_SRCS)
ADD_LIBRARY(botlib ${BOTLIB_SRCS})
TARGET_COMPILE_DEFINITIONS(botlib PRIVATE -DBOTLIB)


AUX_SOURCE_DIRECTORY(code/client CLIENT_SRCS)
IF(NOT USE_CURL)
    list(REMOVE_ITEM CLIENT_SRCS code/client/cl_curl.c)
ENDIF(NOT USE_CURL)
ADD_LIBRARY(client ${CLIENT_SRCS})
TARGET_LINK_LIBRARIES(client qcommon jpeg)
IF(USE_SYSTEM_JPEG)
	TARGET_COMPILE_DEFINITIONS(client PRIVATE -DUSE_SYSTEM_JPEG)
ENDIF(USE_SYSTEM_JPEG)
IF(USE_VULKAN)
	TARGET_COMPILE_DEFINITIONS(client PRIVATE -DUSE_VULKAN_API)
ENDIF(USE_VULKAN)


AUX_SOURCE_DIRECTORY(code/renderercommon RENDERER_COMMON_SRCS)
ADD_LIBRARY(renderer_common ${RENDERER_COMMON_SRCS})
TARGET_LINK_LIBRARIES(renderer_common qcommon)

AUX_SOURCE_DIRECTORY(code/renderer RENDERER_SRCS)
ADD_LIBRARY(renderer ${RENDERER_SRCS})
TARGET_LINK_LIBRARIES(renderer renderer_common)


AUX_SOURCE_DIRECTORY(code/renderervk RENDERER_VK_SRCS)
ADD_LIBRARY(renderer_vk ${RENDERER_VK_SRCS})
TARGET_LINK_LIBRARIES(renderer_vk renderer_common)


IF(USE_SDL)
	AUX_SOURCE_DIRECTORY(code/sdl SDL_SRCS)
	ADD_LIBRARY(qsdl ${SDL_SRCS})
    find_package(SDL2 REQUIRED)
	TARGET_LINK_LIBRARIES(qsdl SDL2::Main)
	IF(USE_VULKAN)
		TARGET_COMPILE_DEFINITIONS(qsdl PRIVATE -DUSE_VULKAN_API)
	ENDIF(USE_VULKAN)

ENDIF()

IF(UNIX)
	SET(Q3_SRCS
            code/unix/unix_main.c
            code/unix/unix_shared.c
            code/unix/linux_signals.c)

	IF (NOT USE_SDL)
		SET(Q3_UI_SRCS
                code/unix/linux_glimp.c
                code/unix/linux_qgl.c
                code/unix/linux_snd.c
                code/unix/x11_dga.c
                code/unix/x11_randr.c
                code/unix/x11_vidmode.c)
		IF(USE_VULKAN_API)
            list(APPEND Q3_UI_SRCS code/unix/linux_qvk.c)
		ENDIF(USE_VULKAN_API)
	ENDIF(NOT USE_SDL)

	IF(APPLE)
		SET(EXE_TYPE MACOSX_BUNDLE)
        list(APPEND Q3_SRCS code/unix/quake3_flat.icns)
        set_source_files_properties(code/unix/quake3_flat.icns PROPERTIES
                MACOSX_PACKAGE_LOCATION Resources)
		SET(MACOSX_BUNDLE_ICON_FILE quake3_flat.icns)
	ENDIF()
ELSEIF(WIN32)
	SET(EXE_TYPE WIN32)
	SET(Q3_SRCS
            code/win32/win_main.c
            code/win32/win_shared.c
            code/win32/win_syscon.c
            code/win32/win_resource.rc)
	IF (NOT USE_SDL)
		SET(Q3_UI_SRCS
                code/win32/win_gamma.c
                code/win32/win_glimp.c
                code/win32/win_input.c
                code/win32/win_minimize.c
                code/win32/win_qgl.c
                code/win32/win_snd.c
                code/win32/win_wndproc.c)
		IF(USE_VULKAN_API)
            list(APPEND Q3_UI_SRCS code/win32/win_qvk.c)
		ENDIF(USE_VULKAN_API)
	ENDIF(NOT USE_SDL)
ENDIF(UNIX)


IF(CMAKE_SYSTEM_PROCESSOR MATCHES AMD|x86)
	SET(VM_SRCS code/qcommon/vm_x86.c)
ELSEIF(CMAKE_SYSTEM_PROCESSOR MATCHES aarch|arm64)
	SET(VM_SRCS code/qcommon/vm_aarch64.c)
ELSEIF(CMAKE_SYSTEM_PROCESSOR MATCHES arm)
	SET(VM_SRCS code/qcommon/vm_armv7l.c)
ELSE()
	TARGET_COMPILE_DEFINITIONS(qcommon PRIVATE -DNO_VM_COMPILED)
ENDIF()

ADD_EXECUTABLE(quake3e ${EXE_TYPE} ${VM_SRCS} ${Q3_SRCS} ${Q3_UI_SRCS})
TARGET_LINK_LIBRARIES(quake3e client botlib m dl)
TARGET_LINK_OPTIONS(quake3e PRIVATE -no-pie)


IF(USE_SDL)
	TARGET_LINK_LIBRARIES(quake3e qsdl)
ENDIF(USE_SDL)

IF(USE_VULKAN)
	TARGET_LINK_LIBRARIES(quake3e renderer_vk)
	TARGET_COMPILE_DEFINITIONS(quake3e PRIVATE -DUSE_VULKAN_API)
ELSE()
	TARGET_LINK_LIBRARIES(quake3e renderer)
ENDIF(USE_VULKAN)

ADD_EXECUTABLE(quake3e.ded ${EXE_TYPE} ${VM_SRCS} ${Q3_SRCS})
TARGET_LINK_LIBRARIES(quake3e.ded qcommon_ded botlib)
TARGET_LINK_OPTIONS(quake3e.ded PRIVATE -no-pie)

IF(WIN32)
	TARGET_LINK_LIBRARIES(quake3e winmm comctl32 ws2_32)
	TARGET_LINK_LIBRARIES(quake3e-server winmm comctl32 ws2_32)
ENDIF(WIN32)
```

**Comment by Alex20129** [comment]
with this new `CMakeLists.txt` file, i get an error:

```
unix_shared.c.o: undefined reference to symbol dlclose@@GLIBC_2.2.5
```

then i added `dl` in 142th line

```TARGET_LINK_LIBRARIES(quake3e client botlib m dl)```

and now it's Ok!

```
[100%] Linking C static library librenderer.a
[100%] Built target renderer
alex@Alex-WS:~/Quake3e/build$
```

Compilation result (quake3e file) looks like 'shared library' on my system, but it works well, i can run it as normal executable. Please, add `-no-pie` option in `TARGET_LINK_OPTIONS`.

**Comment by Alex20129** [comment]
Fixed in #159
